### DISCUSSION TITLE
Render without parent component?

### DISCUSSION THREAD

**Discussion by honzajavorek** [jpsca/jinjax#18] [Q&A]
Hi! 👋 I think jinjax is something I've been looking for for a very long time. Now as I'm learning how to start with it, I figured out the starting point of rendering is that I tell the catalogue which top-level component I wish to render. But what if there's no top-level component? 🤔 My use cases would be:

1. I'd like to drop components inside Markdown, effectively achieving something like [MDX](https://mdxjs.com/). The Markdown file lives e.g. in MkDocs and it's a plain Markdown text except of a few places where I want to include a component. I can pre-process such file with Jinja (MkDocs have a `on_page_markdown` hook), but I cannot give it a top-level component (perhaps I can, but it wouldn't be practical).
2. I have a rather large project where I'd like to start using jinjax. It's a mess of Jinja macros and whatnot. There are, of course, no top-level components at the moment. I cannot go and rewrite everything, I need to roll the change gradually. I need the old code to live alongside the new code. In such environment, I need the standard `template.render(**context)` to be able to go through the Jinja code and process it as always, but also be able to process a component if it's present inside a normal Jinja code.

Is something like that possible, or are we talking about mutually exclusive worlds? Is it either all jinjax or nothing?

**Comment by jpsca** [comment]
Hi! 👋 

Answering your second question:

JinjaX can definitely be used to render only a section of a page. To do so, you call `catalog.render()` not for a single top-level component, but for each component you need to insert in your traditionally-rendered templates, and each of those will render their own sub-components as usual/

Let me elaborate.

1. First, create a catalog as shown in https://jinjax.scaletti.dev/guide/#usage.
2. Second, sdd the catalog as a 'global' variable so it is accessible from your current templates.
3. Then, you can either:

a)  and then make the call from inside the traditional template. For example: `{{ catalog.render('Settings', data=data) }}` or `{{ catalog.render('Button', foo=bar, disabled=False) }}`; or

b) Do the same in your view and pass the resulting HTML as a variable to standard template:
```python
def my_view():
    ...
    footer = catalog.render('Footer')
    return render('my_page.html', footer=footer, **more_context)
```

Note that the standard templates doesn't even need to be Jinja-based

**Reply by honzajavorek** [reply]
Nice! This didn't occur to me. This is pretty flexible.

**Comment by jpsca** [comment]
Almost forgot: you also need to insert the CSS and JS urls of the components in the template that render the `<head>` of the page. Because you've added the `catalog` as global  to your templates, you can do:

`base.html`

```html+jinja
<html>
<head>
   ...
   {{ catalog.render_assets() }}}
</head>
<body>
{% block 'body' %}{% endblock %}
</body>
</html>
```

AFAIK Jinja renders the "child" templates first and then the wrapper component, so even if you use the (a) option, at that point the catalog will have a list of all the asets... but if not, the (b) option will work for sure

**Reply by honzajavorek** [reply]
Thanks!

**Comment by jpsca** [comment]
About your first question... it can present a few challenges.... and I know because I've tried: JinjaX documentation it's actually rendered with a very alpha-stage/experimental system like that: https://github.com/jpsca/ClayDocs

The trickiest one is dealing with source code in the markdown: you need to wrap any embedded code in the markdown with `{% raw %} ... {% endraw %}`, otherwise there is a risk that code contains tags that JinjaX interprets as components, for example, 
```markdown
The `<Button>` components are awesome!
```
will either fail because of a lack of `<Button>` component, or worse, render it when you just wanted to write the text `"<Button>"`.

To deal with that I render the markdown first, and then replacing each `"<code"` with `"{% raw %}<code"`, and each `"</code>"` with `"</code>{% endraw %}"`, an _then_ render the result with JinjaX _as a string_  using the secret attribute `__source`:

```python
catalog.render("RandomName", __source="...source of the page here...'")`
```

**Reply by honzajavorek** [reply]
Wouldn't it be possible to write a Python-Markdown extension, which would identify the tags in the right context (i.e. ignoring them when inside code block, etc.) and render them using the catalog? I'm not an expert on such extensions, but seems like the right thing to do to me, as in the ideal case, the Markdown parser should know everything it needs to decide when to render and when not. An inspiration could be [this article](https://alberand.com/markdown-custom-tags.html), which I only scanned, but which seems to have the right keywords 😄

**Reply by honzajavorek** [reply]
(And by "wouldn't it be possible...?" I mean I could try to contribute, but first I want to know what do you think about such approach.)

**Reply by jpsca** [reply]
That's a excellent idea, and yes I think it would work much better like that than my approach
